### ISSUE TITLE
Add two Iterable methods

### ISSUE THREAD

**Issue by nate-thegrate** [dart-lang/sdk#55037]
```dart
class Iterable<E> {
  E? firstWhereOrNull(bool Function(E) test) {
    try {
      return firstWhere(test);
    } on StateError {
      return null;
    }
  }

  Iterable<(E, T)> zip<T>(Iterable<T> other) sync* {
    final Iterator<E> iterator = this.iterator;
    final Iterator<T> otherIterator = other.iterator;

    do {
      yield (iterator.current, otherIterator.current);
    } while (iterator.moveNext() && otherIterator.moveNext());
  }
}
```

<br>

### firstWhereOrNull

To get the first item in an iterable, you can use `.first` (throws a `StateError` if it's empty) or `.firstOrNull` (returns `null` if it's empty).

The same concept will apply to `.firstWhere()` and `.firstWhereOrNull()` once this method is added.

<br>

### zip

Thanks to Record types, the `zip()` method that already exists in a few other languages is now achievable in Dart.

<br><br>

This issue isn't very urgent, since anyone can use these methods by making an `extension on Iterable`. But due to their prevalent use cases, I think these two methods should be part of the core Dart SDK.

**Comment by lrhn** [comment]
The `firstWhereOrNull` method already exist in `package:collection`: [`firstWhereOrNull`](https://pub.dev/documentation/collection/latest/collection/IterableExtension/firstWhereOrNull.html).

A record based `zip` is an option.
I vividly remember programming it at least once, but it might never have reached maturity. 

With zip, there is always the problem of what to do if the iterables have different lengths. The only *good* solution is to stop, because the alternative is to introduce `null`s.
But the `zip` is really just:
```dart
extension RecordZip2<T1, T2> on (Iterable<T1>, Iterable<T2>) {
  Iterable<(T1, T2)> get zip sync* {
    var it1 = $1.iterator, it2 = $2.iterator;
    while (it1.moveNext() && it2.moveNext()) yield (it1.current, it2.current);
  }
}
```
That's something we could add to `package:collection` when we release a Dart 3.0 version of it, and something we can fairly easily add to `dart:core` too.

**Comment by Reprevise** [comment]
I think `firstWhereOrNull` could definitely be added to `dart:collection` just as `firstOrNull` was in `IterableExtensions`.

Having `zip` as an extension on a record of iterables is less desirable then having it inside of `IterableExtensions` IMO, just because of discoverability.

**Comment by rneswold** [comment]
Did this make it into 3.0? I can't seem to find it in the documentation.

**Comment by nate-thegrate** [comment]
As mentioned previously, [`firstWhereOrNull`](https://pub.dev/documentation/collection/latest/collection/IterableExtension/firstWhereOrNull.html) is already part of the collection package.

To my knowledge, the ability to zip iterables is not yet implemented. If/when it gets added, I'll make sure to close this issue!

**Comment by Reprevise** [comment]
I think firstWhereOrNull is such a mainstream method it should be in the core library.

**Comment by rneswold** [comment]
Yes, `firstWhereOrNull` is in the `IterablesExtension` package of the SDK. But `.zip()` and `.unzip()` aren't.

Day 1 of 2024's Advent of Code had a need for zipping and unzipping. I was surprised Dart's `Iterable` API didn't have them. I used the above `extension` example to make them, but I really feel they should both be in the collection package.

**Comment by rneswold** [comment]
`IterableZip` zips together the values into a list so they all have to be the same type. A more general zip would create a record of any types.

```dart
extension Zip<T1, T2> on Iterable<T1> {
  Iterable<(T1, T2)> zip(Iterable<T2> ii) sync* {
    var it1 = iterator, it2 = ii.iterator;

    while (it1.moveNext() && it2.moveNext()) {
      yield (it1.current, it2.current);
    }
  }
}
```

so

```dart
void main() {
  final fruits = ['Apple', 'Cherry', 'Blueberry'];
  final totals = [1000, 200, 50];
  final zipped = fruits.zip(totals);
  zipped.forEach(print); // (Apple, 1000)
                         // (Cherry, 200)
                         // (Blueberry, 50)
}
```

**Comment by warcayac** [comment]
@rneswold Your extension can be improved like this:

```dart
extension Zip<T1> on Iterable<T1> {
  Iterable<(T1, T2)> zip<T2>(Iterable<T2> other) sync* {
    final it1 = iterator;
    final it2 = other.iterator;

    while (it1.moveNext() && it2.moveNext()) {
      yield (it1.current, it2.current);
    }
  }
}
```

With this small change you ensure that you get the correct type when the two lists have different types, otherwise the second element of each resulting record would be inferred as dynamic.
